src/lib/data/langs/langs.js

```diff
@@ -1,11 +1,46 @@
 import LangsList from '@/lib/data/langs/langs-list.json'
 import L10nSingleton from '@/lib/l10n/l10n-singleton.js'
 
+let allLangs = []

```

Comment:
I agree that this code could be rather confusing, but let me give you some explanations:

this class has the following actions:

- it uploads once a big json file with available languages
- while it uploads this json file - it collects `allLangs` array for options configuration and `rtlLangs` array for direction definition; we collect them once and use several times in the code

`collectLangsData` - uploads languges and creates `allLangs` and `rtlLangs`
`all` - returns `allLangs` 
`defineDirection` - returns check based on `rtlLangs`

That's why - I check if these arrays are ready before a method will try to return the value

As for scoped variables - I like to use such an approach with static classes - that don't have an instance.
This way - one instance of the variable would be created and used. And such a static lass could be easily requested in any place in the code. I am not sure how it is correct in english - but I think of it as a `single instance class`

Reply:
Thanks for explaining this to me, it makes more sense now. My question is: Don't you get the same functionality with [static properties](https://javascript.info/static-properties-methods)? I thought that's how static properties behaved, but I'm not an expert in JavaScript classes, so I could be mistaken.

---

> I check if these arrays are ready before a method will try to return the value

One design pattern that may be useful is putting the check in the collect method instead of in the part of the code that calls the method.

Instead of:

```javascript
static get all () {
  if (allLangs.length === 0) {
    allLangs = this.collectLangsData()
  }
  return allLangs
}

static defineDirection (langCode) {
  if (rtlLangs.length === 0) {
    allLangs = this.collectLangsData()
  }
  return rtlLangs.includes(langCode) ? 'rtl' : 'ltr'
}

static collectLangsData () {
  ...
}
```

You could write it like this:

```javascript
static get all () {
  return this.collectLangsData()
}

static defineDirection (langCode) {
  return collectRtlLangs().includes(langCode) ? 'rtl' : 'ltr'
}

static collectLangsData () {
  if (allLangs.length !== 0) { return allLangs }
  ...
}

static collectRtLangsData () {
  if (rtLangs.length !== 0) { return rtLangs }
  ...
}
```

The benefit of doing it this way is that the you limit the methods that need to be aware of the state, making the code easier to reason about, and making it simpler to add additional functionality.

Reply:
Thank you for the advice - I made similiar refactoring